cpp/spglib.hpp

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include <array>
+#include <exception>
+#include <memory>
+#include <string>
+#include <tuple>
+#include <vector>
+
+namespace Spglib {
+	/**
+	 * Get the version of the Spglib C++ binding
+	 *
+	 * @return X.Y.Z version
+	 */
+	const std::string version();
+	/**
+	 * Get the full version of the Spglib C++ binding
+	 *
+	 * @return full version
+	 */
+	const std::string version_full();
+	/**
+	 * Get the commit of the Spglib C++ binding
+	 *
+	 * @return commit
+	 */
+	const std::string commit();
+	/**
+	 * C++ exception object for spglib related errors
+	 */
+	class SpglibError : public std::exception {
+		const std::string errorMsg;
+
+	public:
+		const char* what() const noexcept override;
+		SpglibError(std::string&& message);
+	};
+	class Symmetry {
+		// TODO: Find better interface compatible with std::mdspan
+		std::array<double, 9> rotations;
+		std::array<double, 3> translations;
+	};
+	class MagneticSymmetry : public Symmetry {
+		bool time_reversal;
+	};
+	class Atom {
+		std::array<double, 3> position;
+		int type;
+	};
+	// TODO: Better name?
+	class Lattice {
+		std::array<double, 9> lattice;
+		std::vector<Atom> atoms;
+		// TODO: Allow for non-periodic dimensions
+	};
+	class PrimitiveLattice : public Lattice {
+		Lattice& original_lattice;
+	};
+	class Spacegroup {
+		int number;
+		int hall_number;
+		std::string international_symbol;
+		std::string hall_symbol;
+		std::string schoenflies;
+		std::string pointgroup_international;
+		std::string pointgroup_schoenflies;
+		int arithmetic_crystal_class_number;
+		std::string arithmetic_crystal_class_symbol;
+		// TODO: Unclear what this is referencing
+		char choice[6];
+	};
+	// TODO: should it inherit/reference Spacegroup
+	class MagneticSpacegroup {
+		int number;
+		int type;
+		int uni_number;
+		int litvin_number;
+		std::string bns_number;
+		std::string og_number;
+	};
```

Comment:
Isn't the `MagneticSpacegroup` also a `Spacegroup`?

Reply:
Magnetic space group and space group are different groups and corresponding objects live in different spaces.

Similar to space group, `MagneticSpacegroupType` is more appropriate.

Reply:
But from what I see in `SpglibMagneticDataset` it also has a `hall_number` associated. So do all `MagneticSpacegroupType` have a `SpacegroupType`/`hall_number`/etc. associated with it?

Reply:
Hall symbols are defined for space group type with selected crystallographic settings. A subset of each magnetic space group (MSG) forms a space group, and the Hall symbol of the space group is useful information in MSG. But we have to remember that the MSG and the subset that forms space group in general don't have group-subgroup relation.

Reply:
Hmm, we could store it as a pointer on MSG, pointing to the SG